File: lib/java/src/main/java/io/github/brenbar/uapi/internal/validation/ValidateNumber.java
package io.github.brenbar.uapi.internal.validation;

import java.math.BigDecimal;
import java.math.BigInteger;
import java.util.List;
import java.util.Map;

import static io.github.brenbar.uapi.internal.types.UNumber._NUMBER_NAME;
import static io.github.brenbar.uapi.internal.validation.GetTypeUnexpectedValidationFailure.getTypeUnexpectedValidationFailure;

public class ValidateNumber {
    public static List<ValidationFailure> validateNumber(Object value) {
        if (value instanceof BigInteger bi || value instanceof BigDecimal bd) {
            return List.of(
                    new ValidationFailure(List.of(), "NumberOutOfRange", Map.of()));
        } else if (value instanceof Number) {
            return List.of();
        } else {
            return getTypeUnexpectedValidationFailure(List.of(), value, _NUMBER_NAME);
        }
    }
}

File: lib/ts/uapi/internal/validation/ValidateNumber.ts
import { get_type_unexpected_validation_failure } from 'uapi/internal/validation/GetTypeUnexpectedValidationFailure';
import { ValidationFailure } from 'uapi/internal/validation/ValidationFailure';
import { _NUMBER_NAME } from 'uapi/internal/types/UNumber';

export function validateNumber(value: any): ValidationFailure[] {
    if (typeof value === 'number' && !isNaN(value) && !isNaN(parseFloat(value))) {
        if (Number.isInteger(value)) {
            if (value > 2 ** 63 - 1 || value < -(2 ** 63)) {
                return [new ValidationFailure([], 'NumberOutOfRange', {})];
            } else {
                return [];
            }
        } else {
            return [];
        }
    } else {
        return get_type_unexpected_validation_failure([], value, _NUMBER_NAME);
    }
}
